File: backend/src/main/java/com/zennyt/games/domain/vo/ScoreBreakdown.java
package com.zennyt.games.domain.vo;

import java.util.List;

/**
 * Détail du calcul du score, sous forme de lignes « comme des logs » — pour que
 * le joueur comprenne d'où viennent ses points.
 *
 * <p><b>Calculé côté serveur</b> à partir des mêmes métriques et du même barème
 * que le score (aucune nouvelle règle) ; le client ne fait que l'afficher. Le
 * mock mobile reproduit ces lignes à l'identique (parité).
 *
 * <p>Chaque ligne porte un {@link Kind} qui indique comment l'afficher :
 * <ul>
 *   <li>{@code NOTE} — phrase d'explication (detail seul) ;</li>
 *   <li>{@code INFO} — mesure informative « label : detail » (sans points) ;</li>
 *   <li>{@code CRITERION} — critère noté « label (detail) : points/max » ;</li>
 *   <li>{@code SUBTOTAL} — sous-total d'un niveau « label = points/max » ;</li>
 *   <li>{@code TOTAL} — total/moyenne « label : points/max ».</li>
 * </ul>
 */
public record ScoreBreakdown(List<Line> lines) {

    public ScoreBreakdown {
        lines = List.copyOf(lines);
    }

    public enum Kind { NOTE, INFO, CRITERION, SUBTOTAL, TOTAL }

    public record Line(Kind kind, String label, String detail, Integer points, Integer maxPoints) {

        public static Line note(String text) {
            return new Line(Kind.NOTE, text, null, null, null);
        }

        public static Line info(String label, String detail) {
            return new Line(Kind.INFO, label, detail, null, null);
        }

        public static Line criterion(String label, String detail, int points, int maxPoints) {
            return new Line(Kind.CRITERION, label, detail, points, maxPoints);
        }

        public static Line subtotal(String label, int points, int maxPoints) {
            return new Line(Kind.SUBTOTAL, label, null, points, maxPoints);
        }

        public static Line total(String label, int points, int maxPoints) {
            return new Line(Kind.TOTAL, label, null, points, maxPoints);
        }
    }
}

File: mobile/lib/features/games/domain/entities/score_breakdown.dart
/// Type d'une ligne du détail du score. Aligné sur ScoreBreakdownLine.kind
/// du contrat games.openapi.yaml.
enum ScoreBreakdownKind { note, info, criterion, subtotal, total }

/// Une ligne du DÉTAIL du score (panneau « d'où viennent mes points »).
///
/// Calculée côté serveur (ou mock) à partir des mêmes métriques et du même
/// barème que le score — le client ne fait que l'afficher.
class ScoreBreakdownLine {
  const ScoreBreakdownLine({
    required this.kind,
    required this.label,
    this.detail,
    this.points,
    this.maxPoints,
  });

  final ScoreBreakdownKind kind;
  final String label;
  final String? detail;
  final int? points;
  final int? maxPoints;

  factory ScoreBreakdownLine.fromJson(Map<String, dynamic> json) {
    return ScoreBreakdownLine(
      kind: _kindFromWire(json['kind'] as String?),
      label: (json['label'] as String?) ?? '',
      detail: json['detail'] as String?,
      points: (json['points'] as num?)?.toInt(),
      maxPoints: (json['maxPoints'] as num?)?.toInt(),
    );
  }

  static ScoreBreakdownKind _kindFromWire(String? wire) {
    switch (wire) {
      case 'NOTE':
        return ScoreBreakdownKind.note;
      case 'INFO':
        return ScoreBreakdownKind.info;
      case 'CRITERION':
        return ScoreBreakdownKind.criterion;
      case 'SUBTOTAL':
        return ScoreBreakdownKind.subtotal;
      case 'TOTAL':
        return ScoreBreakdownKind.total;
      default:
        return ScoreBreakdownKind.info;
    }
  }
}
